### PR TITLE
[#197] Keep the template project up to date with new configurations, versions from the Flutter starter project

### PR DESCRIPTION
**Pull request by luongvo**
- Solves #197

## What happened 👀

- Android
  - Update Kotlin to 1.7.10
  - Update Gradle to 7.3
  - Update Manifest structure
  - [io.flutter.embedding.android.SplashScreenDrawable is now deprecated](https://github.com/flutter/flutter/issues/93668)
- iOS
  - Update default app icons
  - `MinimumOSVersion` to 11
  - Fix: build configurations for both `production` and `staging` schemes. Now we have 4 build configurations for each:
    - Staging Debug
    - Staging Release
    - Production Debug
    - Production Release
  - Update `project.pbxproj` to latest project structure.
- Flutter base
  - Remove redundant `cupertino_icons`
  - Update some dependencies to the latest versions
  - ~Add `useMaterial3: true,` to base `ThemeData` in `main.dart`~

## Insight 📝

- [flutter_config does not work with the newest Gradle version](https://github.com/ByneappLLC/flutter_config/issues/55). While waiting for the official support, we have to switch to the alternative https://pub.dev/packages/flutter_config_plus ✅ . Updating to `flutter_config_plus` requires:
  - Android's `minSdkVersion 21` 👉 `23`
  - iOS's `MinimumOSVersion` to 11
- After updating `permission_handler: ^10.2.0`, we must run this command to let Mason re-generates the bundle file ✅ 

	```
	mason bundle bricks/permission_handler -t dart -o bricks/template/hooks/bundles
	```

## Proof Of Work 📹

| Env | Staging | Production |
|--------|--------|--------|
| Android | Screenshot_20230717-183503 | Screenshot_20230717-184518 |
| iOS | Simulator Screenshot - iPhone 14 Pro Max - 2023-07-17 at 18 55 21 | Simulator Screenshot - iPhone 14 Pro Max - 2023-07-17 at 21 59 38 |

### CODE REVIEW

File: bricks/template/__brick__/{{project_name.snakeCase()}}/.gitignore [hunk 1]
```diff
@@ -1,41 +1,7 @@
...
-/lib/bundler/man/
-
-# for a library or gem, you might want to ignore these files since the code is
-# intended to run in multiple environments; otherwise, check them in:
-# Gemfile.lock
-# .ruby-version
-# .ruby-gemset
-
-# unless supporting rvm < 1.11.0 or doing something fancy, ignore this:
-.rvmrc
```
Comment (luongvo) [inline bricks/template/__brick__/{{project_name.snakeCase()}}/.gitignore:37]: FYI: we don't need them & some from above for Flutter project

File: bricks/template/__brick__/{{project_name.snakeCase()}}/.gitignore [hunk 2]
```diff
@@ -88,9 +55,6 @@ app.*.map.json
 /android/app/profile
 /android/app/release
 
-# Python
-.venv/
```
Comment (luongvo) [inline bricks/template/__brick__/{{project_name.snakeCase()}}/.gitignore:92]: FYI: no more Python

File: bricks/template/__brick__/{{project_name.snakeCase()}}/android/app/build.gradle [hunk 3]
```diff
@@ -35,22 +35,27 @@ project.ext.envConfigFiles = [
...
+    }
 
     sourceSets {
         main.java.srcDirs += 'src/main/kotlin'
-        production {
-            res.srcDirs = ['src/production/res']
-        }
-        staging {
-            res.srcDirs = ['src/staging/res']
-        }
```
Comment (luongvo) [inline bricks/template/__brick__/{{project_name.snakeCase()}}/android/app/build.gradle:47]: FYI: the Android sourceSets for the default `res` folder is redundant

**Review by doannimble — APPROVED** [review]

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -41,7 +41,7 @@ jobs:
         run: flutter packages pub run build_runner build --delete-conflicting-outputs
 
       - name: Check for any formatting issues in the code
-        run: flutter format --set-exit-if-changed .
+        run: dart format --set-exit-if-changed .
```
Comment (luongvo) [inline .github/workflows/test.yml:44]: FYI: This command upgrade is required. However, upgrading to use `dart run` is not required yet s,o we will do it later to not keep too much things in this PR 🙏

**Comment by luongvo** [comment]
flutter_config seems back to work properly in version 2.0.2 https://github.com/ByneappLLC/flutter_config/commits/master 🤔

**Comment by luongvo** [comment]
@doannimble I answered your concern [here](https://github.com/nimblehq/flutter-templates/commit/94f5fb1d62b3d6d306639395bc3ee8472a411d86#r121573246).

File: bricks/template/__brick__/{{project_name.snakeCase()}}/ios/Podfile.lock [hunk 5]
```diff
@@ -6,15 +6,15 @@ PODS:
     - Flutter
   - package_info_plus (0.4.5):
     - Flutter
-  - "permission_handler (5.1.0+2)":
+  - permission_handler_apple (9.0.4):
```
Comment (manh-t) [inline bricks/template/__brick__/{{project_name.snakeCase()}}/ios/Podfile.lock:9]: @luongvo `permission_handler` is already removed from the brick `template`. Can you help to update this file to get rid of `permission_handler` stuff? 🙏

Comment (luongvo) [reply bricks/template/__brick__/{{project_name.snakeCase()}}/ios/Podfile.lock:9]: @manh-t good point, I cherry-picked it from a sample project with permission_handler enabled.

**Review by doannimble — APPROVED** [review]
LGTM now 🚀

File: bricks/template/__brick__/{{project_name.snakeCase()}}/ios/Podfile.lock (hunk 5 above)
Comment (luongvo) [reply bricks/template/__brick__/{{project_name.snakeCase()}}/ios/Podfile.lock]: Fixed in https://github.com/nimblehq/flutter-templates/pull/201/commits/5f5d1e25b12fa3b69b01bb632d55e2c6fe9604e2

**Review by chornerman — COMMENTED** [review]
@luongvo I'm curious do we have a plan to do this with the sample folder? 🤔

**Comment by luongvo** [comment]
@chornerman We have this workflow https://github.com/nimblehq/flutter-templates/blob/develop/.github/workflows/generate_sample.yml 😄

**Comment by chornerman** [comment]
@luongvo I see now, thanks for the explanation 😄

File: bricks/template/__brick__/{{project_name.snakeCase()}}/lib/main.dart [hunk 6]
```diff
@@ -41,6 +41,8 @@ class MyApp extends StatelessWidget {
         primarySwatch: Colors.blue,
         brightness: Brightness.light,
         fontFamily: Assets.fonts.neuzeit,
+        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
+        useMaterial3: true,
```
Comment (sleepylee) [inline bricks/template/__brick__/{{project_name.snakeCase()}}/lib/main.dart:45]: I'm not sure if we should use material3 as default - maybe you would want a feature flag to enable it or not when init the project. Besides, this sounds like out-of-scope for the changes you are doing, so I recommend to separate this config in a different PR. Yes - it is small, but it has a big impact.

**Comment by doannimble** [comment]
@luongvo Just a curious question, should we update the `pubspec.yaml` of sample too? 🤔

**Comment by luongvo** [comment]
@doannimble I have the same answer here, please check it out https://github.com/nimblehq/flutter-templates/pull/201#issuecomment-1635564057 👋

File: bricks/template/__brick__/{{project_name.snakeCase()}}/lib/main.dart (hunk 6 above)
Comment (luongvo) [reply bricks/template/__brick__/{{project_name.snakeCase()}}/lib/main.dart:45]: @sleepylee That makes sense. I think we can create a new feature ticket for material3 support with some updates from Theming by @manh-t 🙆‍♂️

Comment (luongvo) [reply bricks/template/__brick__/{{project_name.snakeCase()}}/lib/main.dart]: @sleepylee Remove in https://github.com/nimblehq/flutter-templates/pull/201/commits/108474daa897512fed11601b87d062861aa7d832. Material3 support will be handled in the new task #236

**Comment by luongvo** [comment]
POW is updated (without Material3 enabled).

File: bricks/template/__brick__/{{project_name.snakeCase()}}/lib/main.dart (hunk 6 above)
Comment (manh-t) [reply bricks/template/__brick__/{{project_name.snakeCase()}}/lib/main.dart]: @luongvo I will create the ticket and work on that 🙆‍♂️

Comment (luongvo) [reply bricks/template/__brick__/{{project_name.snakeCase()}}/lib/main.dart]: @manh-t let me assign you to this task https://github.com/nimblehq/flutter-templates/issues/236 🙆‍♂️

**Review by markgravity — APPROVED** [review]

**Review by kaungkhantsoe — APPROVED** [review]

**Comment by luongvo** [comment]
@manh-t @sleepylee your final reviews are appreciated 🤗

**Review by manh-t — APPROVED** [review]
LGTM now 🙆‍♂️
